pkg/web/apps/apps_repository.go

```diff
@@ -7,4 +7,5 @@ import (
 // Repository represent the app's repository contract
 type Repository interface {
 	GetApps() (*[]models.App, error)
+	GetAppVersionsByAppID(id string) (*[]models.Version, error)
```

Comment:
Should not be here return GetAppsByID() which indeed will be part of /apps/{id} router and task and not this one?

Reply:
@camilamacedo86 `GetAppVersionsByAppID` is required for the `/apps` endpoint. @austincunningham is working on the `GetAppsByID` endpoint

Reply:
Hi @craicoverflow,

What is the struct that will return in the endpoint? Then, you don't need to make queries to return this structure no matter what tables each will be looking for. You just need one query which will return all data required which will be used to create the ResponseObject. 

Important: I understand that you need to return in the endpoint apps/ is the app struct.

Reply:
Hi @camilamacedo86 ,

I believe it not possible with the standard SQL library in Go to map the apps table and the joined table (version) into a Go struct in a single query. It is possible to retrieve the data with SQL, the limitation is in **mapping it to Go**.

Reply:
Hi @craicoverflow,

Sorry, but I do not understand what is the limitation here. Should something as the following example not work here?

```go
func TestQuery(t *testing.T) {
    /* . . . */
    rows, err := db.Query("SELECT 
                a.id ID,
               a.app_id AppID,
               a.app_name AppName,
	      COUNT(v.id) as num_of_deployed_versions NumOfDeployedVersions, 
	      COALESCE(SUM(v.num_of_app_launches),0) as num_of_app_launches NumOfClients,
	       COALESCE(SUM(v.num_of_clients),0) as num_of_clients NumOfAppLaunches
	FROM app as a 
	LEFT JOIN version as v on a.app_id = v.app_id 
	WHERE a.deleted_at IS NULL 
	GROUP BY a.id;`)")
    if err != nil {
            t.Fatalf("Query: %v", err)
    }

    type row struct {
            id  int
            AppID string
           // TODO the rest of impl.  
    }   

    got := []row{}
    for app.Next() {
            var r row
            err = rows.Scan(&r.id, &r.AppID, &r.AppName, &r.NumOfDeployedVersions, &r.NumOfClients, &r.NumOfAppLaunches )
            if err != nil {
                    t.Fatalf("Scan: %v", err)
            }
            got = append(got, r)
    }
    /* . . . */
} 
```

Reply:
```go
func (a *appsPostgreSQLRepository) GetApps() (*[]models.App, error) {
	rows, err := a.db.Query(`
	SELECT a.id,a.app_id,a.app_name,
	COALESCE(COUNT(DISTINCT v.id),0) as num_of_deployed_versions,
	COALESCE(SUM(DISTINCT v.num_of_app_launches),0) as num_of_app_launches,
	COALESCE(SUM(DISTINCT v.num_of_clients),0) as num_of_clients
	FROM app as a LEFT JOIN version as v on a.app_id = v.app_id 
	WHERE a.deleted_at IS NULL 
	GROUP BY a.id;`)

	if err != nil {
		log.Error(err)
		return nil, models.ErrInternalServerError
	}

	defer func() {
		err := rows.Close()
		if err != nil {
			log.Error(err)
		}
	}()

	apps := []models.App{}
	for rows.Next() {
		var a models.App
		if err = rows.Scan(&a.ID, &a.AppID, &a.AppName, &a.NumOfDeployedVersions, &a.NumOfAppLaunches, &a.NumOfClients); err != nil {
			log.Error(err)
		}

		apps = append(apps, a)
	}

	if len(apps) == 0 {
		return nil, models.ErrNotFound
	}

	return &apps, nil
}
```

Apart from the fact that your above example is in a test function, what is different?

Reply:
So, what is not possible?

Reply:
The original comment in this thread is regarding `GetAppVersionsByAppID` so I believed you were talking about something different. We are no longer talking about `GetAppVersionsByAppID`.

Reply:
I am was. I do not think it should be here in this PR since its not part of this task. However, it is fine. I will close this comment because shows that it lost the focus with so misunderstands.

Reply:
The data is required for the `/apps` endpoint so it is a requirement for this PR.